File: go/test/folder_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/typebot-sdk/go"
	"github.com/voxgig-sdk/typebot-sdk/go/core"

	vs "github.com/voxgig-sdk/typebot-sdk/go/utility/struct"
)

func TestFolderEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Folder(nil)
		if ent == nil {
			t.Fatal("expected non-nil FolderEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"folder": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Folder(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.MakeConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Folder(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := folderBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "folder." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set TYPEBOT_TEST_FOLDER_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		folderRef01Ent := client.Folder(nil)
		folderRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "folder"}, setup.data), "folder_ref01"))

		folderRef01DataResult, err := folderRef01Ent.Create(folderRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		folderRef01Data = core.ToMapAny(folderRef01DataResult)
		if folderRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if folderRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		folderRef01Match := map[string]any{}

		folderRef01ListResult, err := folderRef01Ent.List(folderRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		folderRef01List, folderRef01ListOk := folderRef01ListResult.([]any)
		if !folderRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", folderRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(folderRef01List), map[string]any{"id": folderRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		folderRef01DataUp0Up := map[string]any{
			"id": folderRef01Data["id"],
		}

		folderRef01MarkdefUp0Name := "created_at"
		folderRef01MarkdefUp0Value := fmt.Sprintf("Mark01-folder_ref01_%d", setup.now)
		folderRef01DataUp0Up[folderRef01MarkdefUp0Name] = folderRef01MarkdefUp0Value

		folderRef01ResdataUp0Result, err := folderRef01Ent.Update(folderRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		folderRef01ResdataUp0 := core.ToMapAny(folderRef01ResdataUp0Result)
		if folderRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if folderRef01ResdataUp0["id"] != folderRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if folderRef01ResdataUp0[folderRef01MarkdefUp0Name] != folderRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", folderRef01MarkdefUp0Name, folderRef01ResdataUp0[folderRef01MarkdefUp0Name])
		}

		// LOAD
		folderRef01MatchDt0 := map[string]any{
			"id": folderRef01Data["id"],
		}
		folderRef01DataDt0Loaded, err := folderRef01Ent.Load(folderRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		folderRef01DataDt0LoadResult := core.ToMapAny(folderRef01DataDt0Loaded)
		if folderRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if folderRef01DataDt0LoadResult["id"] != folderRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		folderRef01MatchRm0 := map[string]any{
			"id": folderRef01Data["id"],
		}
		_, err = folderRef01Ent.Remove(folderRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		folderRef01MatchRt0 := map[string]any{}

		folderRef01ListRt0Result, err := folderRef01Ent.List(folderRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		folderRef01ListRt0, folderRef01ListRt0Ok := folderRef01ListRt0Result.([]any)
		if !folderRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", folderRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(folderRef01ListRt0), map[string]any{"id": folderRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func folderBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "folder", "FolderTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read folder test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse folder test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"folder01", "folder02", "folder03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("TYPEBOT_TEST_FOLDER_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"TYPEBOT_TEST_FOLDER_ENTID": idmap,
		"TYPEBOT_TEST_LIVE":      "FALSE",
		"TYPEBOT_TEST_EXPLAIN":   "FALSE",
		"TYPEBOT_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["TYPEBOT_TEST_FOLDER_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["TYPEBOT_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["TYPEBOT_APIKEY"],
			},
			extra,
		})
		client = sdk.NewTypebotSDK(core.ToMapAny(mergedOpts))
	}

	live := env["TYPEBOT_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["TYPEBOT_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/FolderEntityTest.php
<?php
declare(strict_types=1);

// Folder entity test

require_once __DIR__ . '/../typebot_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class FolderEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = TypebotSDK::test(null, null);
        $ent = $testsdk->Folder(null);
        $this->assertNotNull($ent);
    }

    // Feature #4: the entity stream(action, ...) method runs the op pipeline
    // and yields result items. With the streaming feature active it yields the
    // feature's incremental output; otherwise it falls back to the materialised
    // list so stream always yields.
    public function test_stream(): void
    {
        $seed = [
            "entity" => [
                "folder" => [
                    "s1" => ["id" => "s1"],
                    "s2" => ["id" => "s2"],
                    "s3" => ["id" => "s3"],
                ],
            ],
        ];

        // Fallback: streaming inactive -> yields the materialised list items.
        $base = TypebotSDK::test($seed, null);
        $seen = iterator_to_array($base->Folder(null)->stream("list", null, null), false);
        $this->assertCount(3, $seen);

        // Inbound: streaming active -> yields each item from the feature.
        $cfg = TypebotConfig::make_config();
        if (isset($cfg["feature"]) && is_array($cfg["feature"]) && isset($cfg["feature"]["streaming"])) {
            $sdk = TypebotSDK::test($seed, ["feature" => ["streaming" => ["active" => true]]]);
            $got = [];
            foreach ($sdk->Folder(null)->stream("list", null, null) as $item) {
                if (is_array($item) && array_is_list($item)) {
                    foreach ($item as $sub) {
                        $got[] = $sub;
                    }
                } else {
                    $got[] = $item;
                }
            }
            $this->assertCount(3, $got);
        }
    }

    public function test_basic_flow(): void
    {
        $setup = folder_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "list", "update", "load", "remove"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "folder." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set TYPEBOT_TEST_FOLDER_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $folder_ref01_ent = $client->Folder(null);
        $folder_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.folder"), "folder_ref01"));

        $folder_ref01_data_result = $folder_ref01_ent->create($folder_ref01_data, null);
        $folder_ref01_data = Helpers::to_map($folder_ref01_data_result);
        $this->assertNotNull($folder_ref01_data);
        $this->assertNotNull($folder_ref01_data["id"]);

        // LIST
        $folder_ref01_match = [];

        $folder_ref01_list_result = $folder_ref01_ent->list($folder_ref01_match, null);
        $this->assertIsArray($folder_ref01_list_result);

        $found_item = sdk_select(
            Runner::entity_list_to_data($folder_ref01_list_result),
            ["id" => $folder_ref01_data["id"]]);
        $this->assertNotEmpty($found_item);

        // UPDATE
        $folder_ref01_data_up0_up = [
            "id" => $folder_ref01_data["id"],
        ];

        $folder_ref01_markdef_up0_name = "created_at";
        $folder_ref01_markdef_up0_value = "Mark01-folder_ref01_" . $setup["now"];
        $folder_ref01_data_up0_up[$folder_ref01_markdef_up0_name] = $folder_ref01_markdef_up0_value;

        $folder_ref01_resdata_up0_result = $folder_ref01_ent->update($folder_ref01_data_up0_up, null);
        $folder_ref01_resdata_up0 = Helpers::to_map($folder_ref01_resdata_up0_result);
        $this->assertNotNull($folder_ref01_resdata_up0);
        $this->assertEquals($folder_ref01_resdata_up0["id"], $folder_ref01_data_up0_up["id"]);
        $this->assertEquals($folder_ref01_resdata_up0[$folder_ref01_markdef_up0_name], $folder_ref01_markdef_up0_value);

        // LOAD
        $folder_ref01_match_dt0 = [
            "id" => $folder_ref01_data["id"],
        ];
        $folder_ref01_data_dt0_loaded = $folder_ref01_ent->load($folder_ref01_match_dt0, null);
        $folder_ref01_data_dt0_load_result = Helpers::to_map($folder_ref01_data_dt0_loaded);
        $this->assertNotNull($folder_ref01_data_dt0_load_result);
        $this->assertEquals($folder_ref01_data_dt0_load_result["id"], $folder_ref01_data["id"]);

        // REMOVE
        $folder_ref01_match_rm0 = [
            "id" => $folder_ref01_data["id"],
        ];
        $folder_ref01_ent->remove($folder_ref01_match_rm0, null);

        // LIST
        $folder_ref01_match_rt0 = [];

        $folder_ref01_list_rt0_result = $folder_ref01_ent->list($folder_ref01_match_rt0, null);
        $this->assertIsArray($folder_ref01_list_rt0_result);

        $not_found_item = sdk_select(
            Runner::entity_list_to_data($folder_ref01_list_rt0_result),
            ["id" => $folder_ref01_data["id"]]);
        $this->assertEmpty($not_found_item);

    }
}

function folder_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/folder/FolderTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = TypebotSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["folder01", "folder02", "folder03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("TYPEBOT_TEST_FOLDER_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "TYPEBOT_TEST_FOLDER_ENTID" => $idmap,
        "TYPEBOT_TEST_LIVE" => "FALSE",
        "TYPEBOT_TEST_EXPLAIN" => "FALSE",
        "TYPEBOT_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["TYPEBOT_TEST_FOLDER_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["TYPEBOT_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["TYPEBOT_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new TypebotSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["TYPEBOT_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["TYPEBOT_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
